prov/sm2/src/sm2_progress.c

```diff
@@ -201,11 +202,143 @@ static int sm2_progress_recv_msg(struct sm2_ep *ep,
 	return ret < 0 ? ret : 0;
 }
 
+static void sm2_do_atomic(void *src, void *dst, void *cmp,
+			  enum fi_datatype datatype, enum fi_op op, size_t cnt,
+			  uint32_t op_flags)
+{
+	char tmp_result[SM2_ATOMIC_INJECT_SIZE];
+
+	if (ofi_atomic_isswap_op(op)) {
+		ofi_atomic_swap_handler(op, datatype, dst, src, cmp, tmp_result,
+					cnt);
+	} else if (op_flags & SM2_RMA_REQ && ofi_atomic_isreadwrite_op(op)) {
+		ofi_atomic_readwrite_handler(op, datatype, dst, src, tmp_result,
+					     cnt);
+	} else if (ofi_atomic_iswrite_op(op)) {
+		ofi_atomic_write_handler(op, datatype, dst, src, cnt);
+	} else {
+		FI_WARN(&sm2_prov, FI_LOG_EP_DATA,
+			"invalid atomic operation\n");
+	}
+
+	if (op_flags & SM2_RMA_REQ)
+		memcpy(src, op == FI_ATOMIC_READ ? dst : tmp_result,
+		       cnt * ofi_datatype_size(datatype));
+}
+
+static int sm2_progress_inject_atomic(struct sm2_xfer_entry *xfer_entry,
+				      struct fi_ioc *ioc, size_t ioc_count,
+				      size_t *len, struct sm2_ep *ep)
+{
+	struct sm2_atomic_entry *atomic_entry =
+		(struct sm2_atomic_entry *) xfer_entry->user_data;
+	uint8_t *src, *comp;
+	int i;
+
+	switch (xfer_entry->hdr.op) {
+	case ofi_op_atomic_compare:
+		src = atomic_entry->atomic_data.buf;
+		comp = atomic_entry->atomic_data.comp;
+		break;
+	default:
+		src = atomic_entry->atomic_data.data;
+		comp = NULL;
+		break;
+	}
+
+	for (i = *len = 0; i < ioc_count && *len < xfer_entry->hdr.size; i++) {
+		sm2_do_atomic(&src[*len], ioc[i].addr,
+			      comp ? &comp[*len] : NULL,
+			      atomic_entry->atomic_hdr.datatype,
+			      atomic_entry->atomic_hdr.atomic_op, ioc[i].count,
+			      xfer_entry->hdr.op_flags);
+		*len += ioc[i].count *
+			ofi_datatype_size(atomic_entry->atomic_hdr.datatype);
+	}
+
+	if (*len != xfer_entry->hdr.size) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL, "recv truncated");
+		return -FI_ETRUNC;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int sm2_progress_atomic(struct sm2_ep *ep,
+			       struct sm2_xfer_entry *xfer_entry)
+{
+	struct sm2_atomic_entry *atomic_entry =
+		(struct sm2_atomic_entry *) xfer_entry->user_data;
+	struct sm2_domain *domain = container_of(
+		ep->util_ep.domain, struct sm2_domain, util_domain);
+	struct fi_ioc ioc[SM2_IOV_LIMIT];
+	size_t ioc_count;
+	size_t total_len = 0;
+	int err = 0, ret = 0;
+
+	for (ioc_count = 0; ioc_count < atomic_entry->atomic_hdr.rma_ioc_count;
+	     ioc_count++) {
+		ret = ofi_mr_verify(
+			&domain->util_domain.mr_map,
+			atomic_entry->atomic_hdr.rma_ioc[ioc_count].count *
+				ofi_datatype_size(
+					atomic_entry->atomic_hdr.datatype),
+			(uintptr_t *) &(
+				atomic_entry->atomic_hdr.rma_ioc[ioc_count]
+					.addr),
+			atomic_entry->atomic_hdr.rma_ioc[ioc_count].key,
+			ofi_rx_mr_reg_flags(
+				xfer_entry->hdr.op,
+				atomic_entry->atomic_hdr.atomic_op));
```

Comment:
Two nits:
1. this formatting is terrible.  Can we have `fi_ioc *ioc = &(atomic_entry->atomic_hdr.rma_ioc[ioc_count]).  Then all the names are much shorter and will fit in one line.
2. Why not do this in sm2_progress_inject_atomic right before we do the operation?  is it because it's atomic and it all has to be done or not done (I don't think so)?
3. `ioc_count` seems misleading, maybe ioc_index or just `i`?